From bd9664f35c08c85856f65d1bf508d3c8451eef17 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Wed, 10 Sep 2025 14:41:44 +0200
Subject: [PATCH] imgtool: improve key password support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Imgtool can sign a binary for multiple boards, which requires the use of
multiple key pairs. To simplify the automation of this process, a key
password argument is added.

The patch also add support to password retries.

Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
---
 scripts/imgtool/main.py | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/scripts/imgtool/main.py b/scripts/imgtool/main.py
index e24c9a08..237512d3 100755
--- a/scripts/imgtool/main.py
+++ b/scripts/imgtool/main.py
@@ -86,13 +86,17 @@ def save_signature(sigfile, sig):
         f.write(signature)
 
 
-def load_key(keyfile):
+def load_key(keyfile, passwd=None):
     # TODO: better handling of invalid pass-phrase
-    key = keys.load(keyfile)
+    if passwd is not None and isinstance(passwd, str):
+        passwd = passwd.encode('utf-8')
+    key = keys.load(keyfile, passwd)
     if key is not None:
         return key
-    passwd = getpass.getpass("Enter key passphrase: ").encode('utf-8')
-    return keys.load(keyfile, passwd)
+    if passwd is None:
+        passwd = getpass.getpass("Enter key passphrase: ").encode('utf-8')
+        return keys.load(keyfile, passwd)
+    return None
 
 
 def get_password():
-- 
2.25.1

